opencensus/stats/aggregation_data.py

```diff
@@ -126,12 +131,17 @@ def __init__(self,
         self._sum_of_sqd_deviations = sum_of_sqd_deviations
         if bounds is None:
             bounds = []
+        else:
+            assert bounds == list(sorted(set(bounds)))
 
         if counts_per_bucket is None:
             counts_per_bucket = [0 for ii in range(len(bounds) + 1)]
-        elif len(counts_per_bucket) != len(bounds) + 1:
-            raise ValueError("counts_per_bucket length does not match bounds "
-                             "length")
+        else:
+            assert all(cc >= 0 for cc in counts_per_bucket)
+            assert len(counts_per_bucket) == len(bounds) + 1
```

Comment:
Asserts aren't guaranteed to run in production, since some people might run Python without assertions, but we always raise ValueErrors or AssertionErrors for these, I think.

Reply:
Sorry, I should have elaborated. Caveat #1 from [this blog post](https://dbader.org/blog/python-assert-tutorial) mentions it. The [`assert`](https://docs.python.org/3/reference/simple_stmts.html#assert) docs also mention it, but not necessarily emphasising how they shouldn't be used.

Reply:
I tried to use asserts in a way consistent with the blog. An assertion failure here should mean a _programming error_ since we only ever expect to call `DistributionAggregationData.__init__` from `DistributionAggregation.__init__`, where we actually do raise a `ValueError` if the user gives us unsorted bounds.

I.e. precondition checks for API methods should raise `Type`/`ValueError`s (and etc.), but precondition checks in internal-only code can assert instead since bad user input should never cause them to fail. The logic doesn't change if the assertions are removed, and they should only fail if there's a bug in the library.

That said, there's no bright line separating the API from the internals here, and we don't lose anything by changing these to `ValueErrors` (other than signaling the difference between expected and unexpected errors). If you've got a strong preference I don't mind changing it.